sub-packages/bionemo-testing/src/bionemo/testing/harnesses/stop_and_go.py

```diff
@@ -14,139 +14,129 @@
 # limitations under the License.
 
 
+import pathlib
+import tempfile
 from abc import ABC, abstractmethod
-from pathlib import Path
-from typing import Any, Callable, Literal, Sequence, TypedDict
+from typing import Dict, Literal, Sequence, Type, TypeVar
 
 import nemo.lightning as nl
+import pytest
 import pytorch_lightning as pl
 from nemo.collections import llm
-from nemo.lightning import nemo_logger, resume
+from nemo.lightning import resume
+from nemo.lightning.nemo_logger import NeMoLogger
 from nemo.lightning.pytorch import callbacks as nl_callbacks
+from nemo.lightning.pytorch.strategies import MegatronStrategy
+from nemo.utils import logging
 
-from bionemo.core import BIONEMO_CACHE_DIR
 from bionemo.testing import testing_callbacks
+from bionemo.testing.harnesses.mode import Mode
 from bionemo.testing.megatron_parallel_state_utils import distributed_model_parallel_state
+from bionemo.testing.torch import recursive_assert_approx_equal
 
 
-__all__: Sequence[str] = (
-    "get_learning_rate",
-    "get_global_step",
-    "StopAndGoHarness",
-    "MetricsFn",
-    "MetricsDict",
-)
+__all__: Sequence[str] = ("StopAndGoHarness",)
 
-MetricsFn = Callable[[pl.Trainer, pl.LightningModule], Any]
-"""A metrics producing function."""
 
+Callback = TypeVar("Callback", bound=pl.Callback)
+CallbackDict = Dict[Mode, Dict[Type[pl.Callback], pl.Callback]]
 
-class MetricsDict(TypedDict):
-    """Default metrics dict."""
 
-    global_step: MetricsFn
-    learning_rate: MetricsFn
+def get_callback(callbacks: CallbackDict, mode: Mode, callback_type: Type[Callback]) -> Callback:
+    """Returns the callback with the given name and mode.
 
-
-def get_learning_rate(trainer: pl.Trainer, model: pl.LightningModule) -> Any:
-    """Returns the learning rate of the model.
+    Convenience function to make type hinting easier.
 
     Args:
-        trainer (pl.Trainer): The PyTorch Lightning trainer.
-        model (pl.LightningModule): The PyTorch Lightning model.
+        callbacks: The dictionary of callbacks.
+        mode: The mode indicating whether to stop or go.
+        callback_type: The type of the callback.
 
     Returns:
-        Any: The learning rate of the model.
+        pl.Callback: The callback with the given name and mode.
     """
-    return trainer.optimizers[0].param_groups[0]["lr"]
-
-
-def get_global_step(trainer: pl.Trainer, model: pl.LightningModule) -> Any:
-    """Returns the global step of the model.
-
-    Args:
-        trainer (pl.Trainer): The PyTorch Lightning trainer.
-        model (pl.LightningModule): The PyTorch Lightning model.
-
-    Returns:
-        Any: The global step of the model.
-    """
-    return trainer.global_step
+    return callbacks[mode][callback_type]  # type: ignore
 
 
 class StopAndGoHarness(ABC):
```

Comment:
Do we need `unittest.TestCase` to make the class methods work?

Reply:
Right, the base class can't start with `Test` or pytest will try to run it 😄 . I can definitely add a bit to the docstring though